File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
import sys

sys.setrecursionlimit(10000)
T = int(input())


dx = [0,0,-1,1]
dy = [1,-1,0,0]
def DFS(x,y):
    visited[y][x] = 1
    for i in range(4):
        nx = dx[i] + x
        ny = dy[i] + y
        if 0 <= nx < M and 0 <= ny < N:
            if visited[ny][nx] == 0 and graph[ny][nx] == 1:
                DFS(nx,ny)    


#testcase
for _ in range(T):
    #input
    # M = 가로 , N = 세로 K = input 
    M,N,K = list(map(int,input().split(" ")))    
    graph = [[0]*(M) for _ in range(N)]
    visited = [[0]*(M) for _ in range(N)]
    cnt  = 0

    for i in range(K):
        X, Y = list(map(int,input().split(" ")))
        graph[Y][X] = 1
    for i in range(N):
        for j in range(M):
            if visited[i][j] == 0 and graph[i][j] ==1:
                DFS(j,i)
                cnt +=1
    print(cnt)

File: 백준/Silver/1012. 유기농 배추/유기농 배추.swift
    let t = Int(readLine()!)!

    let directions = [
        (1, 0), (-1, 0), (0, 1), (0, -1)
    ]

    for _ in 0..<t {
        let MNK = readLine()!.split { $0 == " " }.map { Int($0)! }

        var grid = Array(repeating: Array(repeating: 0, count: MNK[0]), count: MNK[1])
        var ans = 0



        for _ in 0..<MNK[2] {
            let XY = readLine()!.split { $0 == " " }.map { Int($0)! }
            grid[XY[1]][XY[0]] = 1
        }

        for i in 0..<MNK[1] {
            for j in 0..<MNK[0] {
                if grid[i][j] == 1 {
                    dfs(i, j, &grid, MNK[1], MNK[0])
                    ans += 1
                }
            }
        }
        print(ans)
    }

    func dfs(_ y: Int, _ x: Int, _ grid: inout [[Int]], _ maxY: Int, _ maxX: Int) {
        if grid[y][x] == 1 {
            grid[y][x] = 0 // visited 역할
            for (dx, dy) in directions {
                let nx = x + dx
                let ny = y + dy
                if nx >= 0, nx < maxX, ny >= 0, ny < maxY {
                    dfs(ny, nx, &grid, maxY, maxX)
                }
            }
        }
    }
